monai/metrics/froc.py

```diff
@@ -0,0 +1,137 @@
+):
+    """
+    This function is modified from the official evaluation code of
+    `CAMELYON 16 Challenge <https://camelyon16.grand-challenge.org/>`_, and used to distinguish
+    true positive and false positive predictions. A true positive prediction is defined when
+    the detection point is within the annotated ground truth region.
+
+    Args:
+        probs: an array with shape (n,) that represents the probabilities of the detections.
+            Where, n is the number of predicted detections.
+        ycorr: an array with shape (n,) that represents the Y-coordinates of the detections.
+        xcorr: an array with shape (n,) that represents the X-coordinates of the detections.
+        evaluation_mask: the ground truth mask for evaluation.
+        isolated_tumor_cells: a list of labels (of the evaluation mask)
+            that contains Isolated Tumor Cells.
+        image_level: the level at which the evaluation mask is made.
+
+    Returns:
+        fp_probs: an array that contains the probabilities of the false positive detections.
+        tp_probs: an array that contains the probabilities of the True positive detections.
+        num_of_tumors: number of tumors in the image (excluding Isolate Tumor Cells).
+
+    """
+    assert probs.shape == ycorr.shape == xcorr.shape, "the shapes for coordinates and probabilities should be the same."
+
+    if isinstance(probs, torch.Tensor):
+        probs = probs.detach().cpu().numpy()
+    if isinstance(ycorr, torch.Tensor):
+        ycorr = ycorr.detach().cpu().numpy()
+    if isinstance(xcorr, torch.Tensor):
+        xcorr = xcorr.detach().cpu().numpy()
+    if isinstance(evaluation_mask, torch.Tensor):
+        evaluation_mask = evaluation_mask.detach().cpu().numpy()
+
+    if isolated_tumor_cells is None:
+        isolated_tumor_cells = []
+
+    max_label = np.max(evaluation_mask)
+    tp_probs = np.zeros((max_label,), dtype=np.float32)
+
+    ycorr = (ycorr / pow(2, image_level)).astype(int)
+    xcorr = (xcorr / pow(2, image_level)).astype(int)
+
+    hittedlabel = evaluation_mask[ycorr, xcorr]
+    fp_probs = probs[np.where(hittedlabel == 0)]
+    for i in range(1, max_label + 1):
+        if i not in isolated_tumor_cells and i in hittedlabel:
+            tp_probs[i - 1] = probs[np.where(hittedlabel == i)].max()
+
+    num_of_tumors = max_label - len(isolated_tumor_cells)
+    return fp_probs, tp_probs, num_of_tumors
+
+
+def compute_froc_curve_data(
+    fp_probs: Union[np.ndarray, torch.Tensor],
+    tp_probs: Union[np.ndarray, torch.Tensor],
+    num_of_tumors: int,
+    num_of_samples: int,
+):
+    """
+    This function is modified from the official evaluation code of
+    `CAMELYON 16 Challenge <https://camelyon16.grand-challenge.org/>`_, and used to compute
+    the required data for plotting the Free Response Operating Characteristic (FROC) curve.
+
+    Args:
+        fp_probs: an array that contains the probabilities of the false positive detections for all
```

Comment:
I believe it would be clearer to change:
"for all samples that to be computed" --> "for all images under evaluation"